pkg/kv/kvserver/store.go

```diff
@@ -2192,70 +2191,108 @@ func (s *Store) startLeaseRenewer(ctx context.Context) {
+		SpanOpt:  stop.SterileRootSpan,
+	}, func(ctx context.Context) {
+		ctx, cancel := s.stopper.WithCancelOnQuiesce(ctx)
+		defer cancel()
+
+		// The replicas with an active rangefeed. Updated periodically below. Reuse
+		// the rangeIDs slice across runs to minimize allocation.
+		var rangeIDs roachpb.RangeIDSlice
+		updateRangeIDs := func() roachpb.RangeIDSlice {
+			rangeIDs = rangeIDs[:0]
+			s.rangefeedReplicas.Lock()
+			for rangeID := range s.rangefeedReplicas.m {
+				rangeIDs = append(rangeIDs, rangeID)
 			}
-			closedts.SideTransportCloseInterval.SetOnChange(&st.SV, confChanged)
-			RangeFeedRefreshInterval.SetOnChange(&st.SV, confChanged)
+			s.rangefeedReplicas.Unlock()
+			// Sort the range IDs so that we notify them in the same order on each
+			// iteration. With the pacing below, this helps to ensure a consistent
+			// closed timestamp lag for each range. The closedts-rangefeed-updater
+			// refresh loop operates independently of incoming closed timestamp
+			// updates, so this does not favor some replicas over others.
+			sort.Sort(&rangeIDs)
+			return rangeIDs
+		}
 
-			getInterval := func() time.Duration {
-				refresh := RangeFeedRefreshInterval.Get(&st.SV)
-				if refresh != 0 {
-					return refresh
-				}
-				return closedts.SideTransportCloseInterval.Get(&st.SV)
+		timer := timeutil.NewTimer()
+		defer timer.Stop()
+		errInterrupted := errors.New("waiting interrupted")
+		wait := func(ctx context.Context, until time.Time, interrupt <-chan struct{}) error {
+			now := timeutil.Now()
+			if !now.Before(until) {
+				return nil
+			}
+			timer.Reset(until.Sub(now))
+			select {
+			case <-timer.C:
+				timer.Read = true
+				return nil
+			case <-interrupt:
+				return errInterrupted
+			case <-ctx.Done():
+				return ctx.Err()
 			}
+		}
 
-			for {
-				interval := getInterval()
-				if interval > 0 {
-					timer.Reset(interval)
-				} else {
-					// Disable the side-transport.
-					timer.Stop()
-					timer = timeutil.NewTimer()
+		// Repeatedly handle the closed timestamp update for each range ID.
+		//
+		// The `refresh` interval is the target interval between delivering closed
+		// timestamp updates for a single range. This is the target time in which a
+		// single pass of the loop below should complete. The `smear` interval is
+		// the target interval between wake-ups within a single run.
+		//
+		// We smear the work across the refresh interval to avoid waking up all the
+		// replicas' rangefeed processor and registration goroutines in a short
+		// burst. Doing so without pacing can have a severe impact on goroutine
+		// scheduling latency which impacts foreground latency, as tens of thousands
+		// of goroutines can be marked as runnable in a short period of time.
+		//
+		// Pacing works as follows:
+		// - We spread closed timestamp notifications to replicas across the refresh
+		//   interval (default 200ms).
+		// - We smear across replicas at the granularity of smearInterval (e.g. 1ms)
```

Comment:
I thought that we smear replicas/notifications over refresh interval, not the other way round.

Reply:
Tweaked this sentence a bit.